pkg/proxy/winkernel/proxier.go

```diff
@@ -182,12 +182,14 @@ func (ep *endpointsInfo) Cleanup() {
 	// Never delete a Local Endpoint. Local Endpoints are already created by other entities.
 	// Remove only remote endpoints created by this service
 	if ep.refCount <= 0 && !ep.isLocal {
-		klog.V(4).Infof("Removing endpoints for %v, since no one is referencing it", ep)
-		err := ep.hns.deleteEndpoint(ep.hnsID)
-		if err == nil {
-			ep.hnsID = ""
-		} else {
-			klog.Errorf("Endpoint deletion failed for %v: %v", ep.ip, err)
+		if !proxier.IsHnsEndpointStillUsed(ep.hnsID) {
```

Comment:
This unfortunately breaks encapsulation; the endpointsInfo shouldn't be referring back to the proxier object. For example, updateEndpointsMap() could pass down the proxier.isHnsEndpointStillUsed() function to merge() which would then be able to call the function pointer in Cleanup() without referring directly to proxy.

Reply:
The focus should be on setting the ref count correctly. We should know if the endpoint is being used instead of checking every time.